state-sync/inter-component/consensus-notifications/src/lib.rs

```diff
@@ -0,0 +1,424 @@
+        transactions: Vec<Transaction>,
+        reconfiguration_events: Vec<ContractEvent>,
+    ) -> Result<(), Error>;
+
+    /// Notify state sync to synchronize storage to the specified target.
+    async fn sync_to_target(&self, target: LedgerInfoWithSignatures) -> Result<(), Error>;
+}
+
+/// The consensus component responsible for sending notifications and requests to
+/// state sync.
+///
+/// Note: When a ConsensusNotifier instance is created, state sync must take and
+/// listen to the receiver in the corresponding ConsensusNotificationListener.
+#[derive(Debug)]
+pub struct ConsensusNotifier {
+    notification_sender: mpsc::UnboundedSender<ConsensusNotification>,
+
+    /// Timeout for state sync to respond to consensus when handling a commit
+    /// notification.
+    timeout_ms: u64,
+}
+
+impl ConsensusNotifier {
+    /// Returns a new ConsensusNotifier and ConsensusNotificationListener (to be
+    /// used in conjuction with one another).
+    pub fn new(timeout_ms: u64) -> (Self, ConsensusNotificationListener) {
+        let (notification_sender, notification_receiver) = mpsc::unbounded();
+
+        let consensus_synchronizer = ConsensusNotifier {
+            notification_sender,
+            timeout_ms,
+        };
+        let consensus_listener = ConsensusNotificationListener::new(notification_receiver);
+        (consensus_synchronizer, consensus_listener)
+    }
+}
+
+#[async_trait]
+impl ConsensusNotificationSender for ConsensusNotifier {
+    async fn notify_new_commit(
+        &self,
+        transactions: Vec<Transaction>,
+        reconfiguration_events: Vec<ContractEvent>,
+    ) -> Result<(), Error> {
+        // Only send a notification if transactions have been committed
+        if transactions.is_empty() {
+            return Ok(());
+        }
+
+        // Construct a oneshot channel to receive a state sync response
+        let (callback, callback_receiver) = oneshot::channel();
+        let commit_notification =
+            ConsensusNotification::NotifyCommit(ConsensusCommitNotification {
+                transactions,
+                reconfiguration_events,
+                callback,
+            });
+
+        // Send the notification to state sync
+        if let Err(error) = self
+            .notification_sender
+            .clone()
+            .send(commit_notification)
+            .await
+        {
+            return Err(Error::NotificationError(format!(
+                "Failed to notify state sync of committed transactions! Error: {:?}",
+                error
+            )));
+        }
+
+        // Handle any responses or a timeout
+        if let Ok(response) =
+            timeout(Duration::from_millis(self.timeout_ms), callback_receiver).await
+        {
+            match response {
+                Ok(consensus_notification_response) => consensus_notification_response.result,
+                Err(error) => Err(Error::UnexpectedErrorEncountered(format!("{:?}", error))),
```

Comment:
You know, someday we should just make a `debug!` macro, since we do this tons of places.  I wonder if there's a macro for it, since the only thing I could find is a debug and print macro `dbg!`, but that apparently moves the original.

You could alternatively implement display.

Reply:
True -- there's probably a couple nicer ways of dealing with this... For now, I'll add it to my cleanups pile and see if it's worth revisiting this (and the many other places in the code)